### DISCUSSION TITLE
May we combine GodotJS

### DISCUSSION THREAD

**Discussion by nmerget** [godotjs/GodotJS#9] [General]
Heyho @ialex32x,

I stumbled on this repo. 
It seems you try to achive the same like https://github.com/godotjs/javascript.
Maybe we can combine the projects?

[Geequlim](https://github.com/orgs/godotjs/people/Geequlim) started the project 6 years ago as ECMAScript.
I contributed to it last year, but it seems dead. 
Some weeks ago I started to focus on contributing to the repo and we moved forward by creating an [organization](https://github.com/godotjs).

I hadn't looked into your repo in detail, but it seems that you solved a lot of issues already and enabled cool features.
I'm not a C++ Developer, but I started to understand the code in [this](https://github.com/godotjs/javascript).
But it would be great to bundle the efforts and create the best GodotJS.

What do you think?

**Comment by ialex32x** [comment]
Hello @nmerget, it's exciting that you're also interested in integrating JS into Godot! I read the [source](https://github.com/godotjs/javascript) before starting my current project, and it really inspired me. However, I decided to create a new repository because I want to focus on performance and usability by directly basing it on V8. There were too many breaking changes under the hood if directly basing on the `godotjs/javascript` implementation.  
PS: Even though it's based on v8, the static binding (not implemented the current repo, but tested by myself) performance is still not quite on par with GDScript. I'm still working on improving it to get closer to GDScript's performance, especially when it comes to primitive types.

But, it'll be great if there is a way to combine these two JS lang projects into one solution. I'm willing to put in the effort for this.

**Reply by nmerget** [reply]
Hey @ialex32x, I agree, based on what I see so far you choose another approach and without knowing your code in detail, the fact that you can use TS out of the box and chrome as a debugger is awesome.

A lot of [points](https://github.com/godotjs/javascript/issues/132) Geequlim planed to do are already solved in your repo.

In my point of view I invested two week to fix issues of the old implementation, knowing that there will be a lot of effort to have a "stable" 4.x version. But it makes no sense for me to put more effort in the old stuff, if there is already another solution without those problem. 
I don't know how Geequlim would think about it, but in my case I think your repo is the successor. 
We could mark [the old](https://github.com/godotjs/javascript) as deprecated and move forward with your implementation. 
I could support you right away with GitHub actions, documention, examples, npm & TypeScript stuff - and would dig into your code to see if I can support you there. 

Next steps for me would be that I add you to the [organization](https://github.com/godotjs) as an admin. 
So you could move everything there. Then we can deprecate the old repo and would use your current [docs](https://github.com/ialex32x/GodotJS/tree/main/docs) to update [this](https://godotjs.github.io/).
From this point we could create some "roadmap" for the next steps?

I'm excited to hear from you :)

**Reply by ialex32x** [reply]
I'm thrilled to be able to work on it together, and really looking forward to it a lot.
These are the things I've planned to finish before the `1.0` release:
1. Improve the integration of Godot's scripting features. Although it works now, there are various practical issues that cause the ScriptLanguage implementation to produce unexpected results or even fail to work correctly (I'm very new to Godot😝).
2. Add WebBuild support (the performance may not meet requirements).
3. Add QuickJS support (not essential but as an alternative if v8 isn't suitable for certain situations).

It'll last about 2 or 3 months. As for the docs, since the current version is changing quite a bit, I haven't planned to add or modify much yet.

**Reply by nmerget** [reply]
Great to hear that, I invited you to https://github.com/godotjs.

I would start to inspect your current code and would like to integrate https://github.com/ialex32x/GodotJS-Build for a solid CICD.

There is a scons option with `tests=true` to run unit tests like [this](https://github.com/godotjs/javascript/blob/master/tests/test_javascript.h) maybe this is something I can do as well.
